File: HRForz-API-Node/src/shared/schemas/listing.ts
import { z } from 'zod';

export const ListingFilterSchema = z.object({
  sortBy: z.string().default('createdAt'),
  sortOrder: z.enum(['asc', 'desc']).default('desc'),
}).passthrough();

export const PaginationInputSchema = z.object({
  page: z.number().int().min(1).default(1),
  size: z.number().int().min(1).max(100).default(20),
});

export const ListingRequestSchema = z.object({
  filter: ListingFilterSchema.default({}),
  pagination: PaginationInputSchema.default({ page: 1, size: 20 }),
  paginationFlag: z.boolean().default(true),
});

export type ListingFilter = z.infer<typeof ListingFilterSchema>;
export type PaginationInput = z.infer<typeof PaginationInputSchema>;
export type ListingRequest = z.infer<typeof ListingRequestSchema>;

export function getListingParams(request: ListingRequest) {
  const { filter, pagination, paginationFlag } = request;
  return {
    filter,
    page: pagination.page,
    size: pagination.size,
    offset: (pagination.page - 1) * pagination.size,
    limit: pagination.size,
    paginate: paginationFlag,
    sortBy: filter.sortBy,
    sortOrder: filter.sortOrder.toUpperCase() as 'ASC' | 'DESC',
  };
}

File: HRForz-API/app/shared/schemas/listing.py
from pydantic import BaseModel, Field


class ListingFilter(BaseModel):
    """Base filter — all module filters extend this."""

    sortBy: str = Field(default="created_at")
    sortOrder: str = Field(default="desc", pattern="^(asc|desc)$")

    model_config = {"extra": "ignore"}


class PaginationInput(BaseModel):
    page: int = Field(default=1, ge=1, description="1-indexed page number")
    size: int = Field(default=20, ge=1, le=100, description="Records per page")


class ListingRequest(BaseModel):
    filter: ListingFilter
    pagination: PaginationInput = Field(default_factory=PaginationInput)
    paginationFlag: bool = Field(default=True)

    @property
    def page(self) -> int:
        return self.pagination.page

    @property
    def size(self) -> int:
        return self.pagination.size

    @property
    def offset(self) -> int:
        return (self.pagination.page - 1) * self.pagination.size

    @property
    def limit(self) -> int:
        return self.pagination.size


